api/logs/v1/rules_labels_enforcer.go

```diff
@@ -0,0 +1,179 @@
+package http
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/ghodss/yaml"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/prometheus/model/rulefmt"
+)
+
+var errUnknownRulesContentType = errors.New("Unknown rules response content type")
+
+type rulesData struct {
+	RuleGroups []rulefmt.RuleGroup `json:"groups"`
+}
+
+type prometheusRulesResponse struct {
+	Status    string    `json:"status"`
+	Data      rulesData `json:"data"`
+	Error     string    `json:"error"`
+	ErrorType string    `json:"errorType"`
+}
+
+type lokiRulesResponse = map[string][]rulefmt.RuleGroup
+
+type responseLabelFilter struct {
+	http.ResponseWriter
+	body []byte
+}
+
+func (rlf *responseLabelFilter) Write(b []byte) (int, error) {
+	rlf.body = b
+
+	return rlf.ResponseWriter.Write(b)
+}
+
+// WithEnforeFilterLabels returns a middleware that ensure that all returned
+// rules are filtered by labels applied by the user.
+func WithEnforceRuleLabelFilters(logger log.Logger, labelKeys []string) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			level.Debug(logger).Log("msg", "filtering by labels", "keys", fmt.Sprintf("%s", labelKeys))
+
+			if len(labelKeys) == 0 {
+				next.ServeHTTP(w, r)
+				return
+			}
+
+			f := &responseLabelFilter{ResponseWriter: w}
+
+			next.ServeHTTP(f, r)
+
+			matchers := extractMatchers(r, labelKeys)
+
+			contentType := w.Header().Get("Content-Type")
+
+			previous := fmt.Sprint(len(f.body))
+
+			matcherStr := fmt.Sprintf("%v", matchers)
+			level.Debug(logger).Log("msg", "filtering by matchers", "matchers", matcherStr)
+
+			b, err := filterRules(f.body, contentType, matchers)
+			if err != nil {
+				level.Error(logger).Log("msg", err)
+				f.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+
+			current := fmt.Sprint(len(b))
+
+			level.Debug(logger).Log("msg", "filtered rules by", "previous len", previous, "current len", current)
+
+			f.Header().Set("Content-Length", fmt.Sprint(len(b)))
+
+			_, err = f.Write(b)
+			if err != nil {
+				f.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+		})
+	}
+}
+
+func extractMatchers(r *http.Request, l []string) map[string]string {
+	queryParams := r.URL.Query()
+	matchers := map[string]string{}
+	for _, name := range l {
+		value := queryParams.Get(name)
+		if value != "" {
+			matchers[name] = value
+		}
+	}
+
+	return matchers
+}
+
+func filterRules(body []byte, contentType string, matchers map[string]string) ([]byte, error) {
+	switch contentType {
+	case "application/json":
+		var res prometheusRulesResponse
+		err := json.Unmarshal(body, &res)
+		if err != nil {
+			return nil, err
+		}
+
+		res.Data = filterPrometheusRules(res.Data, matchers)
+
+		return json.Marshal(res)
+
+	case "application/yaml":
+		var res lokiRulesResponse
+		err := yaml.Unmarshal(body, &res)
```

Comment:
I don't think this works. You are using "github.com/ghodss/yaml" to unmarshal a Go object that is meant for "gopkg.in/yaml.v2". Have you tested this